broadcast callback on non shared variables

FredLL-Avaiga opened the pull request:
resolves #1207

~renamed current `broadcast_callback` to `broadcast_callback_on_shared`~

Commit:
broadcast callback on non shared variables
resolves #1207

Commit:
update example

FabienLelaquais commented:
I think this PR is also trying to fix something that was a glitch.

@dinhlongviolin1 I need your blessing here.
This `broadcast_callback_on_shared()` to me is useless: the 'shared' state is not known by users, and should actually never been exposed.
What I expect `broadcast_callback()` to do is invoke a user-defined function, potentially with parameters, for every live state.
If then the user invokes `state.assign()` fine, so be it. But let the user decide.
What do you think?

Another important question is: does anyone see a use case for the `_on_shared` version?
Because I don't see any, and I'm willing to drop all this code...

Thanks!

Commit:
remove shared from the broadcast example

dinhlongviolin1 commented:
> I think this PR is also trying to fix something that was a glitch.
> 
> @dinhlongviolin1 I need your blessing here. This `broadcast_callback_on_shared()` to me is useless: the 'shared' state is not known by users, and should actually never been exposed. What I expect `broadcast_callback()` to do is invoke a user-defined function, potentially with parameters, for every live state. If then the user invokes `state.assign()` fine, so be it. But let the user decide. What do you think?
> 
> Another important question is: does anyone see a use case for the `_on_shared` version? Because I don't see any, and I'm willing to drop all this code...
> 
> Thanks!

I think you are right on this one and I agree. We should drop this code as I myself don't see a use case for it. Don't remember why I add it but hopefully it wont affect anything. @FredLL-Avaiga Please help me drop it in this PR 😺

FredLL-Avaiga commented:
> > I think this PR is also trying to fix something that was a glitch.
> > 
> > @dinhlongviolin1 I need your blessing here. This `broadcast_callback_on_shared()` to me is useless: the 'shared' state is not known by users, and should actually never been exposed. What I expect `broadcast_callback()` to do is invoke a user-defined function, potentially with parameters, for every live state. If then the user invokes `state.assign()` fine, so be it. But let the user decide. What do you think?
> > 
> > Another important question is: does anyone see a use case for the `_on_shared` version? Because I don't see any, and I'm willing to drop all this code...
> > 
> > Thanks!
> 
> I think you are right on this one and I agree. We should drop this code as I myself don't see a use case for it. Don't remember why I add it but hopefully it wont affect anything. @FredLL-Avaiga Please help me drop it in this PR 😺 

I will drop it then

Commit:
- remove broadcast_callback_on_shared
- add state to global ctx  to allow use of state in function call used as bound value

FredLL-Avaiga committed:
Merge branch 'develop' into feature/#1207-broadcast-callback

Commit:
remove broadcast_callback_on_shared

FabienLelaquais commented on doc/gui/examples/broadcast.py:
```diff
@@ -75,4 +72,4 @@ def start_or_stop(state):
 Gui.add_shared_variable("button_text")
 
-Gui(page).run()
+Gui(page).run(debug=True)
```
```suggestion
Gui(page).run()
```

FabienLelaquais commented on doc/gui/examples/broadcast.py:
```diff
@@ -59,9 +57,8 @@ def start_or_stop(state):
-    # Because "button_text" is shared, all clients are updated
-    state.button_text = button_texts[1 if thread else 0]
+    # Update button status for all states.
+    broadcast_callback(state.get_gui(), update_state_var, ["button_text", button_texts[1 if thread else 0]])
```
The legacy code (and comment) are appropriate and should not be changed:
we want to demonstrate that a shared variable (defined as such in `Gui.add_shared_variable("button_text")`) can be propagated to all clients in a single (and simple) line of code.

FabienLelaquais commented on doc/gui/examples/broadcast.py:
```diff
@@ -35,17 +35,15 @@
-    # Update all clients
-    state.broadcast("counter", c)
-
+def update_state_var(state: State, var_name: str, val):
```
```suggestion
# Invoked by the timer, for every client
def update_counter(state: State, var_name: str, val):
    # Update the variable value for this client
```

The function name indicates what it does, in the context of the invocation.

FabienLelaquais commented on doc/gui/examples/broadcast.py:
```diff
@@ -35,17 +35,15 @@
         global counter
         counter = counter + 1
-        broadcast_callback(gui, update_counter, [counter])
+        broadcast_callback(gui, update_state_var, ["counter", counter])
```
```suggestion
        broadcast_callback(gui, update_counter, ["counter", counter])
```

FabienLelaquais commented on frontend/taipy-gui/src/components/Taipy/PaginatedTable.tsx:
```diff
@@ -173,7 +173,7 @@ const PaginatedTable = (props: TaipyPaginatedTableProps) => {
                 return [colsOrder, baseColumns, styTt.styles, styTt.tooltips, hNan, filter];
             } catch (e) {
-                console.info("PTable.columns: " + ((e as Error).message || e));
+                console.info("PTable.columns: ", (e as Error).message || e);
```
Because this may be read by programmers, is there any "better" information we could provide instead of the cryptic "PTable"?

FabienLelaquais commented on taipy/gui/gui.py:
```diff
@@ -1490,18 +1491,13 @@ def _call_user_callback(
-        with _broadcast_callback():
-            # Use global scopes for broadcast callbacks
-            return self._call_user_callback(_DataScopes._GLOBAL_ID, user_callback, args, module_context)
+    ) -> t.Dict[str, t.Any]:
```
👍

FabienLelaquais commented on doc/gui/examples/broadcast.py:
```diff
@@ -15,9 +15,9 @@
+# The value is updated for every client using the broadcast_callback() method.
 # The text of the button that starts or stops the thread is updated on every client's browser
-# using a direct assignment of the state property because the variable is declared 'shared'.
+# using the broadcast_callback() method.
```
```suggestion
# using a direct assignment of the state property because the variable is declared 'shared'.
```
Should be the truth.

FabienLelaquais commented on doc/gui/examples/broadcast.py:
```diff
@@ -15,9 +15,9 @@
 # Demonstrate how to share variable values across multiple clients.
 # This application creates a thread that increments a value every few seconds.
-# The value is updated for every client using the state.broadcast() method.
+# The value is updated for every client using the broadcast_callback() method.
```
```suggestion
# The value is updated for every client in a function invoked by broadcast_callback().
```

FabienLelaquais reviewed (changes requested):
I would like to be closer to the legacy example.

Commit:
some of Fab's comments

Commit:
some of Fab's comments

Commit:
Merci les tests

FabienLelaquais committed:
Expose invoke_callback() and broadcast_callback() as Gui methods.

FredLL-Avaiga commented:
can't approve as I am still an assignee but I would.
@FabienLelaquais I thought you wanted to add a `Gui.broadcast(var_name: str, value: t.Any)` that would call `Gui.broadcast_callback` ?

FredLL-Avaiga commented:
I would if you update the branch ;-)

FabienLelaquais commented:
> I would if you update the branch ;-)

Let me confess I completely forgot.
Thanks!

FabienLelaquais committed:
Added Gui.broadcast_change() and gui.broadcast_changes()

FredLL-Avaiga commented on taipy/gui/gui.py:
```diff
@@ -1448,60 +1448,119 @@ def __call_function_with_args(self, **kwargs):
+            for n, v in enumerate(values):
+                state.assign(n, v)
+
+    def broadcast_changes(self, values: dict[str, t.Any]):
```
I would have used kwargs, no ?
A bit less directive maybe but much more pythonic

FredLL-Avaiga commented on taipy/gui/gui.py:
```diff
@@ -1448,60 +1448,119 @@ def __call_function_with_args(self, **kwargs):
+        """
+        self.broadcast_callback(lambda s, n, v: s.assign(n, v), [var_name, value])
+
+    def __broadcast_changes_fn(state: State, values: dict[str, t.Any]) -> None:
```
should be defined as a static method

FabienLelaquais committed:
Fix Gui.broadcast_changes()

FabienLelaquais committed:
Group shared/broadcast tests

FabienLelaquais committed:
Fix tests

FabienLelaquais committed:
Spelling

FabienLelaquais committed:
Fix mypy

FabienLelaquais committed:
Update doc examples

FabienLelaquais committed:
Linter

FabienLelaquais committed:
More linters

FredLL-Avaiga reviewed (commented):
👍

FabienLelaquais reviewed (dismissed):
Looking good!

FabienLelaquais committed:
Fix tests and make linters even more happy

FabienLelaquais committed:
Fix test [2]

FabienLelaquais committed:
Re-re-re-re-re-fix tests

FabienLelaquais reviewed (dismissed):
➕

namnguyen20999 reviewed (dismissed):
LGTM

dinhlongviolin1 reviewed (dismissed):
Amazingly done!

Commit:
avoid impact of invoke callback on current state

FredLL-Avaiga committed:
Merge branch 'develop' into feature/#1207-broadcast-callback

FabienLelaquais reviewed (approved):
👍
